### PR TITLE
Fixing bugs about the worst simple_animal coded in and material tiles.

### LINKED ISSUE #37421: Floorbots can cause infinite tiles
Issue reported from  Round ID: 86972 (/tg/Station Bagil [ENGLISH] [US-EAST] [100% FREE LAG])

Reporting client version: 512

If a floorbot repairs a tile, you can crowbar up two tiles from one spot.

The round before somehow we had INFINITE tiles coming from a spot and we couldn't DELETE the tile with admin delete.

1. Spawn a lizard
2. Give yourself a fireball spell
3. Fire ball the lizard
4. Spawn floorbots and set them to repair tiles
5. drag two floorbots on the same tile
5. crowbar tile.

### LINKED ISSUE #51038: Emagged Floorbots create bugged lattices.
# Round ID:[136746](https://scrubby.melonmesa.com/round/136746)

Actual Behaviour: Emagged floorbots just put a buggy lattice on top of floor tiles
Expected Behaviour: Deconstructing the tile+plating underneath, exposing the tile to space.

### LINKED ISSUE #54159: Infinite floor panel
Issue reported from  Round ID: [147548](https://scrubby.melonmesa.com/round/147548) (/tg/Station Manuel [ENGLISH] [US-WEST] [ROLEPLAY] [99% LAG FREE])

Reporting client version: 513.1532

Delta station

Two panels down directly from the bottom robotics maint door is a floor panel that could be lifted infinite amount of times.

### LINKED ISSUE #54196: Infinite Floortiles
The floorbot can create floortiles without a catwalk.

Removing these creates a floortile item but doesn't remove it from the ground.

Spamming crowbar on it creates a ton of floortiles.

Issue reported from  Round ID: [147690](https://scrubby.melonmesa.com/round/147690) (/tg/Station - Terry [EU] [100% LAG FREE])

Reporting client version: 513.1533

### LINKED ISSUE #54882: Floorbot converts glass floor tiles into normal floortiles
Instead of replacing floortiles with glassfloor tiles floorbot transmutes them into normal floortiles for some reason, even when you take them out of it, they will become normal.
##Reproduction:
Make a floorbot
Insert some glass tiles
Make floorbot to replace with custom tiles/Take the tiles out
You got yourself normal floortiles now out of any custom ones!

### LINKED ISSUE #55906: Custom Floor Tiles on Floor Bots Broken?

### PR DESCRIPTION
**Pull request by Ghommie**
## About The Pull Request
Fixing two nasty issues with floorbots and another also pertaining material tiles.

Oh yea, I borrowed the can_merge proc from #54858 since it was needed for this refactor, so be sure to merge that first.

Edit: Replaced `set_anchored` usage with `TRAIT_IMMOBILIZED` and `move_resist` because mobs should never be `anchored`.

## Why It's Good For The Game
This will close #54196, close #54159, close #54882, close #37421, close #51038, close #55906. 

## Changelog
:cl:
fix: Fixed floorbots applying floor tiles onto already tiled turfs.
fix: Fixed floorbot issues with material tiles.
fix: Fixed emagged floorbots spawning lattice when deconstructing floor into plating.
/:cl:

### CODE REVIEW

**Comment by Cyprex** [comment]
emagged floorbots are also broken, they don#t properly deconstruct tiles

File: code/game/turfs/open/floor/plating.dm [hunk 1]
```diff
@@ -118,7 +107,7 @@
 	ChangeTurf(/turf/open/floor/plating/rust)
 
 /turf/open/floor/plating/make_plating(force = FALSE)
-	return
+	return src
```
Comment (ShizCalev) [inline code/game/turfs/open/floor/plating.dm:110]: this breaks `break_tile_to_plating()` and `remove_tile()`

**Review by ShizCalev — CHANGES_REQUESTED** [review]

**Comment by Ghommie** [comment]
> 
> 
> emagged floorbots are also broken, they don#t properly deconstruct tiles

done.

**Comment by Cyprex** [comment]
I believe this just needed review? Consider reopening if that's true

**Comment by Ghommie** [comment]
NOT AGAIN! DAMNIT!
Could you please reopen and merge it when you can, @ShizCalev ? I did test it and everything is fine.

File: code/game/objects/items/stacks/stack.dm [hunk 2]
```diff
@@ -407,14 +407,17 @@
...
 	if(QDELETED(S) || QDELETED(src) || S == src) //amusingly this can cause a stack to consume itself, let's not allow that.
 		return
 	var/transfer = get_amount()
 	if(S.is_cyborg)
 		transfer = min(transfer, round((S.source.max_energy - S.source.energy) / S.cost))
 	else
-		transfer = min(transfer, S.max_amount - S.amount)
+		transfer = min(transfer, (limit ? limit : S.max_amount) - S.amount)
+	if(limit)
+		transfer = min(limit, transfer)
```
Comment (LemonInTheDark) [inline code/game/objects/items/stacks/stack.dm:420]: ???

File: code/game/objects/items/stacks/tiles/tile_types.dm [hunk 3]
```diff
@@ -421,6 +426,11 @@
 	material_flags = MATERIAL_ADD_PREFIX | MATERIAL_COLOR | MATERIAL_AFFECT_STATISTICS
 	merge_type = /obj/item/stack/tile/material
 
+/obj/item/stack/tile/material/place_tile(turf/open/T)
+	. = ..()
+	var/turf/open/floor/material/F = .
+	F?.set_custom_materials(mats_per_unit)
```
Comment (LemonInTheDark) [inline code/game/objects/items/stacks/tiles/tile_types.dm:432]: why do it one way with lights, and another with mats? pick one

File: code/modules/mob/living/simple_animal/bot/floorbot.dm [hunk 4]
```diff
@@ -53,6 +52,17 @@
...
 
+/mob/living/simple_animal/bot/floorbot/Exited(atom/movable/A, atom/newloc)
+	if(A == tilestack)
+		if(tilestack && tilestack.max_amount < tilestack.amount) //split the stack if it exceeds its normal max_amount
+			var/iterations = round(tilestack.amount/tilestack.max_amount) //round() without second arg floors the value
+			for(var/a in 1 to iterations)
+				if(a == iterations)
+					tilestack.split_stack(null, tilestack.amount - tilestack.max_amount)
+				else
+					tilestack.split_stack(null, tilestack.max_amount)
```
Comment (LemonInTheDark) [inline code/modules/mob/living/simple_animal/bot/floorbot.dm:63]: Does this code exist elsewhere? feels like it should

File: code/modules/mob/living/simple_animal/bot/floorbot.dm [hunk 5]
```diff
@@ -134,6 +146,19 @@
...
+	if(engage)
+		ADD_TRAIT(src, TRAIT_IMMOBILIZED, BUSY_FLOORBOT_TRAIT)
+		move_resist = INFINITY
+		if(change_icon)
+			icon_state = "[toolbox_color]floorbot-c"
+	else
+		REMOVE_TRAIT(src, TRAIT_IMMOBILIZED, BUSY_FLOORBOT_TRAIT)
+		move_resist = initial(move_resist)
+		if(change_icon)
+			update_icon()
```
Comment (LemonInTheDark) [inline code/modules/mob/living/simple_animal/bot/floorbot.dm:160]: why call it toggle_anchor if that's not what it does?

File: code/modules/mob/living/simple_animal/bot/floorbot.dm [hunk 6]
```diff
@@ -234,9 +254,12 @@
...
 			else if(emagged == 2 && isfloorturf(target))
 				var/turf/open/floor/F = target
-				set_anchored(TRUE)
+				toggle_anchor()
 				mode = BOT_REPAIRING
-				F.ReplaceWithLattice()
+				if(isplatingturf(F))
+					F.ReplaceWithLattice()
+				else
+					F.ScrapeAway(flags = CHANGETURF_INHERIT_AIR)
```
Comment (LemonInTheDark) [inline code/modules/mob/living/simple_animal/bot/floorbot.dm:262]: isn't this intended?

**Review by LemonInTheDark — CHANGES_REQUESTED** [review]

**Comment by TemporalOroboros** [comment]
I assume that this also fixes #55906?

File: code/modules/mob/living/simple_animal/bot/floorbot.dm (hunk 6 above)
Comment (Ghommie) [reply code/modules/mob/living/simple_animal/bot/floorbot.dm:262]: Yes, see #51038.

File: code/modules/mob/living/simple_animal/bot/floorbot.dm (hunk 4 above)
Comment (Ghommie) [reply code/modules/mob/living/simple_animal/bot/floorbot.dm:63]: No, it doesn't afaik.

Comment (AnturK) [reply code/modules/mob/living/simple_animal/bot/floorbot.dm:62]: This should probably be on the stack or tile type and just called here.

**Pull request merged**